pypeit/core/datacube.py

```diff
@@ -267,6 +270,228 @@ def dar_correction(wave_arr, coord, obstime, location, pressure, temperature, re
+        omask = mask+smask
+        idx_sum = 0
+        cntmask = (varcube > 0.0) * omask
+        scimask = flxcube * cntmask
+        varmask = varcube * cntmask**2
+        cnt_spec = cntmask.sum(idx_sum) * utils.inverse(omask.sum(idx_sum))
+        nrmcnt = utils.inverse(cnt_spec)
+        box_sciimg = scimask.sum(idx_sum) * nrmcnt
+        box_scivar = varmask.sum(idx_sum) * nrmcnt**2
+        box_sciivar = utils.inverse(box_scivar)
+        # Transpose for optimal
+        box_sciimg = box_sciimg.T
+        box_sciivar = box_sciivar.T
+
+        # Prepare for optimal
+        msgs.info("Starting optimal extraction")
+        thismask = np.ones(box_sciimg.shape, dtype=np.bool)
+        nspec, nspat = thismask.shape[0], thismask.shape[1]
+        slit_left = np.zeros(nspec)
+        slit_right = np.ones(nspec)*(nspat-1)
+        tilts = np.outer(np.linspace(0.0,1.0,nspec), np.ones(nspat))
+        waveimg = np.outer(wave.value, np.ones(nspat))
+        global_sky = np.zeros_like(box_sciimg)
+        # Find objects and then extract
+        sobj = findobj_skymask.objs_in_slit(box_sciimg, thismask, slit_left, slit_right)
+        skysub.local_skysub_extract(box_sciimg, box_sciivar, tilts, waveimg, global_sky, thismask, slit_left,
+                             slit_right, sobj, model_noise=False)
+        opt_flux, opt_var, opt_gpm = sobj.OPT_COUNTS[0,:], sobj.OPT_COUNTS_SIG[0,:]**2, sobj.OPT_MASK[0,:]
+        # Setup the return values
+        ret_flux, ret_var, ret_gpm = opt_flux, opt_var, opt_gpm
+
+        ###  OLD VERSION OF OPTIMAL USING 2D PROFILE
+        # # Generate a mask
+        # mask = (mask==1).astype(np.float)
+        # fitmask = (varcube > 0.0) * mask
+        # numim = flxcube.shape[0]*flxcube.shape[1]
+        # wl_img -= np.sum(wl_img[:, :, np.newaxis] * smask)/np.sum(smask)
+        # model = wl_img[:, :, np.newaxis] * mask
+        # # Optimally extract
+        # msgs.info("Optimally extracting...")
+        # sciimg = (flxcube*mask).reshape((numim, numwave)).T
+        # ivar = utils.inverse((varcube*mask**2).reshape((numim, numwave)).T)
+        # optmask = (fitmask > 0).reshape((numim, numwave)).T
+        # waveimg = np.ones((numwave, numim))  # Just a dummy array - not needed
+        # skyimg = np.zeros((numwave, numim))  # Just a dummy array - not needed
+        # thismask = np.ones((numwave, numim))  # Just a dummy array - not needed
+        # oprof = model.repeat(numwave, axis=2).reshape((numim, numwave)).T
+        # sobj = specobj.SpecObj('IFU', 'DET01', SLITID=0)
+        # extract.extract_optimal(sciimg, ivar, optmask, waveimg, skyimg, thismask, oprof, sobj)
+        # opt_flux, opt_var, opt_gpm = sobj.OPT_COUNTS, sobj.OPT_COUNTS_SIG**2, sobj.OPT_MASK
+        # # Setup the return values
+        # ret_flux, ret_var, ret_gpm = opt_flux, opt_var, opt_gpm
+    else:
+        msgs.error("Unknown extraction method: ", method)
+
+    # Convert to counts/s/A
```

Comment:
I'm not following why this is counts per second per angstrom and not just counts per second. Where have you divided by the wavelength grid?